### ISSUE TITLE
Why subpixel rather than deconv?

### ISSUE THREAD

**Issue by xuxy09** [Tetrachrome/subpixel#33]
It seems like you highly recommend to use subpixel to replace deconv. But as introduced by Wenzhe Shi, subpixel upsampling is identical  to deconv upsampling. So I don't understand why you do this while deconv layer has been well developed in existed deep learning tools.

**Comment by chroneus** [comment]
no, it is 2 different approaches

**Comment by riokt** [comment]
So what is the advantage of using subpixel rather than only deconv?

**Comment by shifang99** [comment]
compared with deconv,  subpixel  can reduce the computational complexity of the overall SR operation.

**Comment by jaretburkett** [comment]
In tensorflow, doing subpixel is way slower. There are so many splits, concats, reshapes involved that it is way more computationally intensive than a deconvolution. I created a deconvolution kernel that does the same thing and runs about 2-3x faster in my tests. It is hard coded for a 2x upsample. Need 4? Run it through twice.

```python
def SubpixelReshuffleLayer(inputs):
    batch_size, rows, cols, in_channels = inputs.get_shape().as_list()
    kernel_filter_size = 2
    out_channels = int(in_channels // 4)
    kernel_shape = [kernel_filter_size, kernel_filter_size, out_channels, in_channels]
    kernel = np.zeros(kernel_shape, np.float32)

    # Build the kernel so that a 4 pixel cluster has each pixel come from a separate channel.
    for c in range(0, out_channels):
        i = 0
        for x, y in itertools.product(range(2), repeat=2):
            kernel[y, x, c, c * 4 + i] = 1
            i += 1

    new_rows, new_cols = int(rows * 2), int(cols * 2)
    new_shape = [batch_size, new_rows, new_cols, out_channels]
    tf_shape = tf.stack(new_shape)
    strides_shape = [1, 2, 2, 1]

    out = tf.nn.conv2d_transpose(inputs, kernel, tf_shape, strides_shape, padding='VALID')

    return out
```

Although, it goes to show you that instead of building the kernel, you could just train it and could possibly get better results. However, I keep the kernel hardcoded in my projects because I believe there is some magic of each output pixel coming from a single input pixel, but who knows.

In short, a stride of (2,2) on a kernel size of (2,2) and having 1/4 the channels on the output vs input will effectively accomplish the same thing but will be trainable, Or if you want to have it work how subpixel works, you can hard code the kernel as I have done above.
